File writing callback

mpowelson opened the pull request:
Adds the file writing callback, two independent python scripts to analyze the files, and a usage example (in glass upright). The most invasive change is the change to the inputs to callback functions. I changed them to accept the entire OptResults object rather than just the x matrix. The plotting callback was updated to match this.

Levi-Armstrong commented on trajopt_sco/src/optimizers.cpp:
```diff
@@ -476,8 +476,8 @@ OptStatus BasicTrustRegionSQP::optimize()
   return retval;
 }
-}
+}  // namespace sco
```
https://github.com/ros-industrial-consortium/trajopt_ros/blob/01421b492743fc10dc651956062b5053259facd8/trajopt_sco/src/optimizers.cpp#L269

I also propose going ahead and cleaning up this class by removing the use `x_` by replacing it with `results_.x`

Levi-Armstrong commented on trajopt/src/file_write_callback.cpp:
```diff
@@ -0,0 +1,139 @@
+
+    // Write costs to file
+    std::vector<double> costs = results.cost_vals;
+    for (auto cost : costs)
```
```suggestion
    for (const auto& cost : costs)
```

Levi-Armstrong commented on trajopt/src/file_write_callback.cpp:
```diff
@@ -0,0 +1,139 @@
+
+    // Write constraints to file
+    std::vector<double> constraints = results.cnt_viols;
+    for (auto constraint : constraints)
```
```suggestion
    for (const auto& constraint : constraints)
```

Levi-Armstrong commented on trajopt/src/file_write_callback.cpp:
```diff
@@ -0,0 +1,139 @@
+
+  // Write cost names
+  vector<CostPtr> costs = prob->getCosts();
+  for (auto cost : costs)
```
```suggestion
  for (const auto& cost : costs)
```

Levi-Armstrong commented on trajopt/src/file_write_callback.cpp:
```diff
@@ -0,0 +1,139 @@
+
+  // Write constraint names
+  vector<ConstraintPtr> cnts = prob->getConstraints();
+  for (auto cnt : cnts)
```
```suggestion
  for (const auto& cnt : cnts)
```

Levi-Armstrong commented on trajopt/src/plot_callback.cpp:
```diff
@@ -16,8 +16,9 @@ void PlotCosts(const tesseract::BasicPlottingPtr plotter,
-               const DblVec& x)
+               const OptResults& results)
 {
+  auto x=results.x;
```
I would recommend just using results.x throughout.

mpowelson commented on trajopt/include/trajopt/file_write_callback.hpp:
```diff
@@ -0,0 +1,12 @@
+                                  Eigen::Isometry3d tcp = Eigen::Isometry3d::Identity(),
+                                  bool angle_axis = false,
+                                  bool degrees = false);
+}
```
- tcp applies a static transform to the pose in line 43
- angle_axis is a switch to output angle axis rotations instead of quaternions (I imagine to help him evaluate the conical constraint)
- degrees switches angle axis to degrees instead of radians

I am fine with simplifying it if you want. Just let me know what you'd like to do.

mpowelson commented:
@Levi-Armstrong I changed file write to more closely match plotting. I realized after we talked that I imagine you wrote it the way you did so that all of the get<value> functions only have to be called when you create the callback rather than each iteration. Now instead of passing in the full prob, I just pass the parts that we need.

Levi-Armstrong commented on trajopt/src/file_write_callback.cpp:
```diff
@@ -0,0 +1,118 @@
+#include <functional>
```
Need to add the license information.

Levi-Armstrong commented on trajopt/scripts/plot_optimization.py:
```diff
@@ -0,0 +1,149 @@
+''' 
```
Need to add license.

Levi-Armstrong commented on trajopt/scripts/traj_compare.py:
```diff
@@ -0,0 +1,129 @@
+import numpy as np
```
Need to add license

Levi-Armstrong commented:
@mpowelson What are your thought about modifying the Plotter class by renaming it something like DebugUtils which has `plot` and `toString`? For this to work we will need to figure out the best way to generate the header string. Maybe `toString(bool header = false)` where if `header = true` it returns the header otherwise returns the constraint/cost data.

Levi-Armstrong commented:
@mpowelson Package Licenses 

- BSD-2-Clause
  - trajopt
  - trajopt_sco
  - trajopt_utils
- Apache 2.0
  - trajopt_tools
  - trajopt_test_support
  - trajopt_examples
  - trajopt_moveit
  - trajopt_moveit_examples

mpowelson commented:
@Levi-Armstrong Added the license to noted files. 

Creating DebugUtils from Plotter seems like a decent idea to me. Then each cost could handle its own header and the filewriter could be more generic.

mpowelson commented:
@Levi-Armstrong Alright. So after our discussion I changed Trajopt to a dual license BSD, Apache 2.0. The original files are unchanged, but I put the Apache 2.0 license on the file write files since they are new.

I also changed the requested variables to const.

Levi-Armstrong commented on trajopt/src/file_write_callback.cpp:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    // Write costs to file
+    std::vector<double> costs = results.cost_vals;
```
```suggestion
    const std::vector<double>& costs = results.cost_vals;
```

Levi-Armstrong commented on trajopt/src/file_write_callback.cpp:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    // Write constraints to file
+    std::vector<double> constraints = results.cnt_viols;
```
```suggestion
    const std::vector<double>& constraints = results.cnt_viols;
```

Levi-Armstrong commented on trajopt/src/file_write_callback.cpp:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  // Write constraint names
+  vector<ConstraintPtr> cnts = prob->getConstraints();
```
```suggestion
  const vector<ConstraintPtr>& cnts = prob->getConstraints();
```

Levi-Armstrong commented on trajopt/src/file_write_callback.cpp:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  // Write cost names
+  vector<CostPtr> costs = prob->getCosts();
```
```suggestion
  cost vector<CostPtr>& costs = prob->getCosts();
```

Levi-Armstrong commented on trajopt/include/trajopt/file_write_callback.hpp:
```diff
@@ -0,0 +1,9 @@
+namespace trajopt
+{
+Optimizer::Callback WriteCallback(std::shared_ptr<std::ofstream> file,
+                                  TrajOptProbPtr prob);
```
```suggestion
                                  const TrajOptProb& prob);
```
Also probably want to update the plot callback to match.

Levi-Armstrong commented:
@mpowelson A few minor changes and then it should be ready to merge. Also lets save the switching the class name to DebugUtils for another PR.
